fix: start/end positions at peak heights

pcolaianni opened the pull request:
What do you think of this fix? There's some "refactoring" involved.

I report the failures after said changes:
```
547: checking PeakShapeMetrics calculatePeakShapeMetrics( const MSChromatogram& chromatogram, const double left, const double right, const double peak_height, const double peak_apex_pos ) const ... 
547:  -  line 564:  TEST_REAL_SIMILAR(psm.width_at_5,0.231263425125414): got 0.15955, expected 0.231263425125414 (absolute: 0.0717134251254137 [1e-05], relative: 1.44947304998692 [1.00001], message: "ratio of numbers is large"
547:  -  line 565:  TEST_REAL_SIMILAR(psm.width_at_10,0.134762234301732): got 0.138366667, expected 0.134762234301732 (absolute: 0.0036044326982681 [1e-05], relative: 1.02674660832795 [1.00001], message: "ratio of numbers is large"
547:  -  line 566:  TEST_REAL_SIMILAR(psm.width_at_50,0.0595791540757924): got 0.0741500000000004, expected 0.0595791540757924 (absolute: 0.014570845924208 [1e-05], relative: 1.24456281983581 [1.00001], message: "ratio of numbers is large"
547:  -  line 567:  TEST_REAL_SIMILAR(psm.start_position_at_5,2.51268515480125): got 2.630066667, expected 2.51268515480125 (absolute: 0.11738151219875 [1e-05], relative: 1.04671556719888 [1.00001], message: "ratio of numbers is large"
547:  -  line 568:  TEST_REAL_SIMILAR(psm.start_position_at_10,2.63222565817823): got 2.64065, expected 2.63222565817823 (absolute: 0.00842434182176977 [1e-05], relative: 1.00320046337805 [1.00001], message: "ratio of numbers is large"
547:  -  line 569:  TEST_REAL_SIMILAR(psm.start_position_at_50,2.65391757114759): got 2.662116667, expected 2.65391757114759 (absolute: 0.00819909585240985 [1e-05], relative: 1.0030894312399 [1.00001], message: "ratio of numbers is large"
547:  -  line 570:  TEST_REAL_SIMILAR(psm.end_position_at_5,2.74394857992666): got 2.789616667, expected 2.74394857992666 (absolute: 0.0456680870733401 [1e-05], relative: 1.01664320075362 [1.00001], message: "ratio of numbers is large"
547:  -  line 571:  TEST_REAL_SIMILAR(psm.end_position_at_10,2.76698789247996): got 2.779016667, expected 2.76698789247996 (absolute: 0.01202877452004 [1e-05], relative: 1.00434724508652 [1.00001], message: "ratio of numbers is large"
547:  -  line 572:  TEST_REAL_SIMILAR(psm.end_position_at_50,2.71349672522338): got 2.736266667, expected 2.71349672522338 (absolute: 0.0227699417766201 [1e-05], relative: 1.00839136512123 [1.00001], message: "ratio of numbers is large"
547:  +  line 573:  TEST_REAL_SIMILAR(psm.total_width,0.540983333): got 0.540983333, expected 0.540983333
547:  -  line 574:  TEST_REAL_SIMILAR(psm.tailing_factor,5.86240177860251): got 2.14352889450752, expected 5.86240177860251 (absolute: 3.71887288409499 [1e-05], relative: 2.73493013955821 [1.00001], message: "ratio of numbers is large"
547:  -  line 575:  TEST_REAL_SIMILAR(psm.asymmetry_factor,0.864593034054243): got 0.856855285811426, expected 0.864593034054243 (absolute: 0.00773774824281726 [1e-05], relative: 1.00903040264902 [1.00001], message: "ratio of numbers is large"
547:  +  line 576:  TEST_REAL_SIMILAR(psm.slope_of_baseline,2454): got 2454, expected 2454
547:  +  line 577:  TEST_REAL_SIMILAR(psm.baseline_delta_2_height,0.00253908735640033): got 0.00253908735640033, expected 0.00253908735640033
547:  +  line 578:  TEST_EQUAL(psm.points_across_baseline,57): got 57, expected 57
547:  +  line 579:  TEST_EQUAL(psm.points_across_half_height,6): got 6, expected 6
547: : failed
547: 
547: checking PeakShapeMetrics calculatePeakShapeMetrics( const MSChromatogram& chromatogram, MSChromatogram::ConstIterator& left, MSChromatogram::ConstIterator& right, const double peak_height, const double peak_apex_pos ) const ... 
547:  -  line 592:  TEST_REAL_SIMILAR(psm.width_at_5,0.231263425125414): got 0.15955, expected 0.231263425125414 (absolute: 0.0717134251254137 [1e-05], relative: 1.44947304998692 [1.00001], message: "ratio of numbers is large"
547:  -  line 593:  TEST_REAL_SIMILAR(psm.width_at_10,0.134762234301732): got 0.138366667, expected 0.134762234301732 (absolute: 0.0036044326982681 [1e-05], relative: 1.02674660832795 [1.00001], message: "ratio of numbers is large"
547:  -  line 594:  TEST_REAL_SIMILAR(psm.width_at_50,0.0595791540757924): got 0.0741500000000004, expected 0.0595791540757924 (absolute: 0.014570845924208 [1e-05], relative: 1.24456281983581 [1.00001], message: "ratio of numbers is large"
547:  -  line 595:  TEST_REAL_SIMILAR(psm.start_position_at_5,2.51268515480125): got 2.630066667, expected 2.51268515480125 (absolute: 0.11738151219875 [1e-05], relative: 1.04671556719888 [1.00001], message: "ratio of numbers is large"
547:  -  line 596:  TEST_REAL_SIMILAR(psm.start_position_at_10,2.63222565817823): got 2.64065, expected 2.63222565817823 (absolute: 0.00842434182176977 [1e-05], relative: 1.00320046337805 [1.00001], message: "ratio of numbers is large"
547:  -  line 597:  TEST_REAL_SIMILAR(psm.start_position_at_50,2.65391757114759): got 2.662116667, expected 2.65391757114759 (absolute: 0.00819909585240985 [1e-05], relative: 1.0030894312399 [1.00001], message: "ratio of numbers is large"
547:  -  line 598:  TEST_REAL_SIMILAR(psm.end_position_at_5,2.74394857992666): got 2.789616667, expected 2.74394857992666 (absolute: 0.0456680870733401 [1e-05], relative: 1.01664320075362 [1.00001], message: "ratio of numbers is large"
547:  -  line 599:  TEST_REAL_SIMILAR(psm.end_position_at_10,2.76698789247996): got 2.779016667, expected 2.76698789247996 (absolute: 0.01202877452004 [1e-05], relative: 1.00434724508652 [1.00001], message: "ratio of numbers is large"
547:  -  line 600:  TEST_REAL_SIMILAR(psm.end_position_at_50,2.71349672522338): got 2.736266667, expected 2.71349672522338 (absolute: 0.0227699417766201 [1e-05], relative: 1.00839136512123 [1.00001], message: "ratio of numbers is large"
547:  +  line 601:  TEST_REAL_SIMILAR(psm.total_width,0.540983333): got 0.540983333, expected 0.540983333
547:  -  line 602:  TEST_REAL_SIMILAR(psm.tailing_factor,5.86240177860251): got 2.14352889450752, expected 5.86240177860251 (absolute: 3.71887288409499 [1e-05], relative: 2.73493013955821 [1.00001], message: "ratio of numbers is large"
547:  -  line 603:  TEST_REAL_SIMILAR(psm.asymmetry_factor,0.864593034054243): got 0.856855285811426, expected 0.864593034054243 (absolute: 0.00773774824281726 [1e-05], relative: 1.00903040264902 [1.00001], message: "ratio of numbers is large"
547:  +  line 604:  TEST_REAL_SIMILAR(psm.slope_of_baseline,2454): got 2454, expected 2454
547:  +  line 605:  TEST_REAL_SIMILAR(psm.baseline_delta_2_height,0.00253908735640033): got 0.00253908735640033, expected 0.00253908735640033
547:  +  line 606:  TEST_EQUAL(psm.points_across_baseline,57): got 57, expected 57
547:  +  line 607:  TEST_EQUAL(psm.points_across_half_height,6): got 6, expected 6
547: : failed
547: 
547: checking PeakShapeMetrics calculatePeakShapeMetrics( const MSSpectrum& spectrum, const double left, const double right, const double peak_height, const double peak_apex_pos ) const ... 
547:  -  line 620:  TEST_REAL_SIMILAR(psm.width_at_5,0.231263425125414): got 0.15955, expected 0.231263425125414 (absolute: 0.0717134251254137 [1e-05], relative: 1.44947304998692 [1.00001], message: "ratio of numbers is large"
547:  -  line 621:  TEST_REAL_SIMILAR(psm.width_at_10,0.134762234301732): got 0.138366667, expected 0.134762234301732 (absolute: 0.0036044326982681 [1e-05], relative: 1.02674660832795 [1.00001], message: "ratio of numbers is large"
547:  -  line 622:  TEST_REAL_SIMILAR(psm.width_at_50,0.0595791540757924): got 0.0741500000000004, expected 0.0595791540757924 (absolute: 0.014570845924208 [1e-05], relative: 1.24456281983581 [1.00001], message: "ratio of numbers is large"
547:  -  line 623:  TEST_REAL_SIMILAR(psm.start_position_at_5,2.51268515480125): got 2.630066667, expected 2.51268515480125 (absolute: 0.11738151219875 [1e-05], relative: 1.04671556719888 [1.00001], message: "ratio of numbers is large"
547:  -  line 624:  TEST_REAL_SIMILAR(psm.start_position_at_10,2.63222565817823): got 2.64065, expected 2.63222565817823 (absolute: 0.00842434182176977 [1e-05], relative: 1.00320046337805 [1.00001], message: "ratio of numbers is large"
547:  -  line 625:  TEST_REAL_SIMILAR(psm.start_position_at_50,2.65391757114759): got 2.662116667, expected 2.65391757114759 (absolute: 0.00819909585240985 [1e-05], relative: 1.0030894312399 [1.00001], message: "ratio of numbers is large"
547:  -  line 626:  TEST_REAL_SIMILAR(psm.end_position_at_5,2.74394857992666): got 2.789616667, expected 2.74394857992666 (absolute: 0.0456680870733401 [1e-05], relative: 1.01664320075362 [1.00001], message: "ratio of numbers is large"
547:  -  line 627:  TEST_REAL_SIMILAR(psm.end_position_at_10,2.76698789247996): got 2.779016667, expected 2.76698789247996 (absolute: 0.01202877452004 [1e-05], relative: 1.00434724508652 [1.00001], message: "ratio of numbers is large"
547:  -  line 628:  TEST_REAL_SIMILAR(psm.end_position_at_50,2.71349672522338): got 2.736266667, expected 2.71349672522338 (absolute: 0.0227699417766201 [1e-05], relative: 1.00839136512123 [1.00001], message: "ratio of numbers is large"
547:  +  line 629:  TEST_REAL_SIMILAR(psm.total_width,0.540983333): got 0.540983333, expected 0.540983333
547:  -  line 630:  TEST_REAL_SIMILAR(psm.tailing_factor,5.86240177860251): got 2.14352889450752, expected 5.86240177860251 (absolute: 3.71887288409499 [1e-05], relative: 2.73493013955821 [1.00001], message: "ratio of numbers is large"
547:  -  line 631:  TEST_REAL_SIMILAR(psm.asymmetry_factor,0.864593034054243): got 0.856855285811426, expected 0.864593034054243 (absolute: 0.00773774824281726 [1e-05], relative: 1.00903040264902 [1.00001], message: "ratio of numbers is large"
547:  +  line 632:  TEST_REAL_SIMILAR(psm.slope_of_baseline,2454): got 2454, expected 2454
547:  +  line 633:  TEST_REAL_SIMILAR(psm.baseline_delta_2_height,0.00253908735640033): got 0.00253908735640033, expected 0.00253908735640033
547:  +  line 634:  TEST_EQUAL(psm.points_across_baseline,57): got 57, expected 57
547:  +  line 635:  TEST_EQUAL(psm.points_across_half_height,6): got 6, expected 6
547: : failed
547: 
547: checking PeakShapeMetrics calculatePeakShapeMetrics( const MSSpectrum& spectrum, MSSpectrum::ConstIterator& left, MSSpectrum::ConstIterator& right, const double peak_height, const double peak_apex_pos ) const ... 
547:  -  line 648:  TEST_REAL_SIMILAR(psm.width_at_5,0.231263425125414): got 0.15955, expected 0.231263425125414 (absolute: 0.0717134251254137 [1e-05], relative: 1.44947304998692 [1.00001], message: "ratio of numbers is large"
547:  -  line 649:  TEST_REAL_SIMILAR(psm.width_at_10,0.134762234301732): got 0.138366667, expected 0.134762234301732 (absolute: 0.0036044326982681 [1e-05], relative: 1.02674660832795 [1.00001], message: "ratio of numbers is large"
547:  -  line 650:  TEST_REAL_SIMILAR(psm.width_at_50,0.0595791540757924): got 0.0741500000000004, expected 0.0595791540757924 (absolute: 0.014570845924208 [1e-05], relative: 1.24456281983581 [1.00001], message: "ratio of numbers is large"
547:  -  line 651:  TEST_REAL_SIMILAR(psm.start_position_at_5,2.51268515480125): got 2.630066667, expected 2.51268515480125 (absolute: 0.11738151219875 [1e-05], relative: 1.04671556719888 [1.00001], message: "ratio of numbers is large"
547:  -  line 652:  TEST_REAL_SIMILAR(psm.start_position_at_10,2.63222565817823): got 2.64065, expected 2.63222565817823 (absolute: 0.00842434182176977 [1e-05], relative: 1.00320046337805 [1.00001], message: "ratio of numbers is large"
547:  -  line 653:  TEST_REAL_SIMILAR(psm.start_position_at_50,2.65391757114759): got 2.662116667, expected 2.65391757114759 (absolute: 0.00819909585240985 [1e-05], relative: 1.0030894312399 [1.00001], message: "ratio of numbers is large"
547:  -  line 654:  TEST_REAL_SIMILAR(psm.end_position_at_5,2.74394857992666): got 2.789616667, expected 2.74394857992666 (absolute: 0.0456680870733401 [1e-05], relative: 1.01664320075362 [1.00001], message: "ratio of numbers is large"
547:  -  line 655:  TEST_REAL_SIMILAR(psm.end_position_at_10,2.76698789247996): got 2.779016667, expected 2.76698789247996 (absolute: 0.01202877452004 [1e-05], relative: 1.00434724508652 [1.00001], message: "ratio of numbers is large"
547:  -  line 656:  TEST_REAL_SIMILAR(psm.end_position_at_50,2.71349672522338): got 2.736266667, expected 2.71349672522338 (absolute: 0.0227699417766201 [1e-05], relative: 1.00839136512123 [1.00001], message: "ratio of numbers is large"
547:  +  line 657:  TEST_REAL_SIMILAR(psm.total_width,0.540983333): got 0.540983333, expected 0.540983333
547:  -  line 658:  TEST_REAL_SIMILAR(psm.tailing_factor,5.86240177860251): got 2.14352889450752, expected 5.86240177860251 (absolute: 3.71887288409499 [1e-05], relative: 2.73493013955821 [1.00001], message: "ratio of numbers is large"
547:  -  line 659:  TEST_REAL_SIMILAR(psm.asymmetry_factor,0.864593034054243): got 0.856855285811426, expected 0.864593034054243 (absolute: 0.00773774824281726 [1e-05], relative: 1.00903040264902 [1.00001], message: "ratio of numbers is large"
547:  +  line 660:  TEST_REAL_SIMILAR(psm.slope_of_baseline,2454): got 2454, expected 2454
547:  +  line 661:  TEST_REAL_SIMILAR(psm.baseline_delta_2_height,0.00253908735640033): got 0.00253908735640033, expected 0.00253908735640033
547:  +  line 662:  TEST_EQUAL(psm.points_across_baseline,57): got 57, expected 57
547:  +  line 663:  TEST_EQUAL(psm.points_across_half_height,6): got 6, expected 6
547: : failed
```

pcolaianni committed:
fix: start/end positions at peak heights

pcolaianni committed:
Merge branch 'develop' into fix/peak_shape_ints_at_positions

dmccloskey commented on src/openms/include/OpenMS/ANALYSIS/OPENSWATH/PeakIntegrator.h:
```diff
@@ -518,6 +518,15 @@ namespace OpenMS
     ) const;
 
+    template <typename PeakContainerConstIteratorT>
+    double findPosAtPerc_(
```
What is "perc"?  I am guessing that it is percent.  It maybe best just to call it "percent" instead of "perc".

pcolaianni replied:
It is percent

dmccloskey commented on src/openms/source/ANALYSIS/OPENSWATH/PeakIntegrator.cpp:
```diff
@@ -354,4 +321,38 @@ namespace OpenMS
+    const bool is_left_half
+  ) const
+  {
+    const double goal_intensity = peak_height * perc;
```
Let's rename this variable "percent_intensity"

dmccloskey commented on src/openms/source/ANALYSIS/OPENSWATH/PeakIntegrator.cpp:
```diff
@@ -354,4 +321,38 @@ namespace OpenMS
+  {
+    const double goal_intensity = peak_height * perc;
+    PeakContainerConstIteratorT closest = is_left_half ? it_begin : it_end - 1; // TODO assuming enough points here
+    if (is_left_half)
```
Good that this covers both left and right peak side search iterations.

dmccloskey commented on src/openms/source/ANALYSIS/OPENSWATH/PeakIntegrator.cpp:
```diff
@@ -354,4 +321,38 @@ namespace OpenMS
+  ) const
+  {
+    const double goal_intensity = peak_height * perc;
+    PeakContainerConstIteratorT closest = is_left_half ? it_begin : it_end - 1; // TODO assuming enough points here
```
I like that you set the default values here and then override them as you search for a more optimal value (if one is available)

dmccloskey reviewed (changes requested):
I think this implementation addresses the concerns that we had.  We are assuming that the peak is convex.  I have not seen any cases yet where it is not, but potentially that could be an issue that would still throw this algorithm off.

If you can add into the method description for `findPosAtPercentage_` the following:

"The method assumes a convex peak.  If 5%, 10%, or 50% peak heights are not found on either side of the peak, the closest left (for left peak height percentages) and closes right (for right peak height percentages) will be used."

dmccloskey commented on src/openms/source/ANALYSIS/OPENSWATH/PeakIntegrator.cpp:
```diff
@@ -354,4 +321,38 @@ namespace OpenMS
+    }
+    else
+    {
+      for (PeakContainerConstIteratorT it = it_end - 1; it != it_begin; --it)
```
This is an important change that we are iterating from right to left for the right side of the peak.  Nice.

dmccloskey commented:
Additional tests:

1) peak without a left/right 5, 10, or 50% point
2) peak with very few points (i.e., 3 points)

We can think about issuing a subsequent PR at a later point in time if we end up finding non-convex peaks.

pcolaianni commented:
Thanks for the review, I'll take care of those points.
Adding:
- [x] Exit loop if optimal position is found. No need to always iterate the entire half of the peak.

pcolaianni committed:
style: improve method and vars names

rename findPosAtPerc_ to findPosAtPeakHeightPercent_
rename perc to percent
rename goal_intensity to percent_intensity

pcolaianni committed:
docs: Add documentation to findPosAtPeakHeightPercent_()

style: fix argument name

pcolaianni committed:
test: update tests results

pcolaianni committed:
test: Test for past (5,10,50)% peak's height points

pcolaianni committed:
test: Add test for chromatogram/spectrum with few points (3 points)

pcolaianni committed:
feat: exit loop when optimal position is found

pcolaianni commented:
@dmccloskey All changes are applied.
You can see the new commits through here: https://github.com/biosustain/OpenMS/pull/120/files/6904b5d857a6bf86935268bba2fde10cb36d24e6..90cf61ab4441722ed7d58ad3b09549d3e1f82087

dmccloskey commented on src/openms/include/OpenMS/ANALYSIS/OPENSWATH/PeakIntegrator.h:
```diff
@@ -518,12 +518,25 @@ namespace OpenMS
+    */
     template <typename PeakContainerConstIteratorT>
-    double findPosAtPerc_(
+    double findPosAtPeakHeightPercent_(
```
Good call adding in the "PeakHeight"

dmccloskey commented on src/openms/source/ANALYSIS/OPENSWATH/PeakIntegrator.cpp:
```diff
@@ -323,35 +323,33 @@ namespace OpenMS
-          closest = it;
-        }
-      }
+      for (
```
This is pretty cool.   Not only does it save a line, but it also breaks the loop when original if statement is no longer met.

dmccloskey commented on src/tests/class_tests/openms/source/PeakIntegrator_test.cpp:
```diff
@@ -97,6 +106,25 @@ MSChromatogram::ConstIterator chrom_right_it = chromatogram.RTEnd(right) - 1;
 MSSpectrum::ConstIterator spec_left_it = spectrum.MZBegin(left);
 MSSpectrum::ConstIterator spec_right_it = spectrum.MZEnd(right) - 1;
 
+MSChromatogram::ConstIterator chrom_left_past_5_it = chromatogram.RTBegin(left_past_5);
```
Thanks for adding this test case in.

dmccloskey commented on src/tests/class_tests/openms/source/PeakIntegrator_test.cpp:
```diff
@@ -561,22 +589,44 @@ START_SECTION(PeakShapeMetrics calculatePeakShapeMetrics(
   TEST_REAL_SIMILAR(psm.baseline_delta_2_height, 0.00253908735640033)
   TEST_EQUAL(psm.points_across_baseline, 57)
   TEST_EQUAL(psm.points_across_half_height, 6)
+  pa = ptr->integratePeak(chromatogram, left_past_5, right_past_5);
```
This should cover all of the fringe cases that I have seen so far during our testing with analytics.
